{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This notebook demonstrates basic usage of BioThings Explorer, an engine for autonomously querying a distributed knowledge graph. BioThings Explorer can answer two classes of queries -- \"PREDICT\" and \"EXPLAIN\".  PREDICT queries are described in [PREDICT_demo.ipynb](PREDICT_demo.ipynb). Here, we describe EXPLAIN queries and how to use BioThings Explorer to execute them.  A more detailed overview of the BioThings Explorer systems is provided in [these slides](https://docs.google.com/presentation/d/1QWQqqQhPD_pzKryh6Wijm4YQswv8pAjleVORCPyJyDE/edit?usp=sharing).\n",
    "\n",
    "EXPLAIN queries are designed to **identify plausible reasoning chains to explain the relationship between two entities**.  For example, in this notebook, we explore the question:\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\"*Why does imatinib have an effect on the treatment of chronic myelogenous leukemia (CML)?*\"  \n",
    "\n",
    "Later, we also compare those results to a similar query looking at imatinib's role in treating gastrointestinal stromal tumors (GIST).\n",
    "\n",
    "**To experiment with an executable version of this notebook, [load it in Google Colaboratory](https://colab.research.google.com/github/biothings/biothings_explorer/blob/master/jupyter%20notebooks/EXPLAIN_demo.ipynb).**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 0: Load BioThings Explorer modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, install the `biothings_explorer` and `biothings_schema` packages, as described in this [README](https://github.com/biothings/biothings_explorer/blob/master/jupyter%20notebooks/README.md#prerequisite).  This only needs to be done once (but including it here for compability with [colab](https://colab.research.google.com/))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!pip install git+https://github.com/biothings/biothings_explorer#egg=biothings_explorer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, import the relevant modules:\n",
    "\n",
    "* **Hint**: Find corresponding bio-entity representation used in BioThings Explorer based on user input (could be any database IDs, symbols, names)\n",
    "* **FindConnection**: Find intermediate bio-entities which connects user specified input and output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules from biothings_explorer\n",
    "from biothings_explorer.hint import Hint\n",
    "from biothings_explorer.user_query_dispatcher import FindConnection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Find representation of \"chronic myelogenous leukemia\" and \"imatinib\" in BTE\n",
    "\n",
    "In this step, BioThings Explorer translates our query strings \"chronic myelogenous leukemia\" and \"imatinib\" into BioThings objects, which contain mappings to many common identifiers.  Generally, the top result returned by the `Hint` module will be the correct item, but you should confirm that using the identifiers shown.\n",
    "\n",
    "Search terms can correspond to any child of [BiologicalEntity](https://biolink.github.io/biolink-model/docs/BiologicalEntity.html) from the [Biolink Model](https://biolink.github.io/biolink-model/docs/), including `DiseaseOrPhenotypicFeature` (e.g., \"lupus\"), `ChemicalSubstance` (e.g., \"acetaminophen\"), `Gene` (e.g., \"CDK2\"), `BiologicalProcess` (e.g., \"T cell differentiation\"), and `Pathway` (e.g., \"Citric acid cycle\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mondo': 'MONDO:0011996',\n",
       " 'doid': 'DOID:8552',\n",
       " 'umls': 'C1292772',\n",
       " 'mesh': 'D015464',\n",
       " 'name': 'chronic myelogenous leukemia',\n",
       " 'display': 'mondo(MONDO:0011996) doid(DOID:8552) umls(C1292772) mesh(D015464) name(chronic myelogenous leukemia) ',\n",
       " 'type': 'DiseaseOrPhenotypicFeature',\n",
       " 'primary': {'identifier': 'mondo',\n",
       "  'cls': 'DiseaseOrPhenotypicFeature',\n",
       "  'value': 'MONDO:0011996'}}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ht = Hint()\n",
    "# find all potential representations of CML\n",
    "cml_hint = ht.query(\"chronic myelogenous leukemia\")\n",
    "# select the correct representation of CML\n",
    "cml = cml_hint['DiseaseOrPhenotypicFeature'][0]\n",
    "cml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chembl': 'CHEMBL941',\n",
       " 'drugbank': 'DB00619',\n",
       " 'name': 'Imatinib',\n",
       " 'pubchem': 5291,\n",
       " 'umls': 'C0935989',\n",
       " 'mesh': 'D000068877',\n",
       " 'chebi': 'CHEBI:45783',\n",
       " 'smiles': 'Cc1ccc(NC(=O)c2ccc(CN3CCN(C)CC3)cc2)cc1Nc1nccc(-c2cccnc2)n1',\n",
       " 'display': 'chembl(CHEMBL941) drugbank(DB00619) name(Imatinib) pubchem(5291) umls(C0935989) mesh(D000068877) chebi(CHEBI:45783) smiles(Cc1ccc(NC(=O)c2ccc(CN3CCN(C)CC3)cc2)cc1Nc1nccc(-c2cccnc2)n1) ',\n",
       " 'type': 'ChemicalSubstance',\n",
       " 'primary': {'identifier': 'chembl',\n",
       "  'cls': 'ChemicalSubstance',\n",
       "  'value': 'CHEMBL941'}}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find all potential representations of imatinib\n",
    "imatinib_hint = ht.query(\"imatinib\")\n",
    "# select the correct representation of imatinib\n",
    "imatinib = imatinib_hint['ChemicalSubstance'][0]\n",
    "imatinib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Find intermediate nodes connecting imatinib and chronic myelogenous leukemia\n",
    "\n",
    "In this section, we find all paths in the knowledge graph that connect imatinib and chronic myelogenous leukemia.  To do that, we will use `FindConnection`.  This class is a convenient wrapper around two advanced functions for **query path planning** and **query path execution**. More advanced features for both query path planning and query path execution are in development and will be documented in the coming months. \n",
    "\n",
    "The parameters for `FindConnection` are described below:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function __init__ in module biothings_explorer.user_query_dispatcher:\n",
      "\n",
      "__init__(self, input_obj, output_obj, intermediate_nodes, registry=None)\n",
      "    Find relationships in the Knowledge Graph between an Input Object and an Output Object.\n",
      "    \n",
      "    Args:\n",
      "        input_obj (required): must be an object returned from Hint corresponding to a specific biomedical entity.\n",
      "                            Examples:\n",
      "                Hint().query(\"Fanconi anemia\")['DiseaseOrPhenotypicFeature'][0]\n",
      "                Hint().query(\"acetaminophen\")['ChemicalSubstance'][0]\n",
      "    \n",
      "        output_obj (required): must EITHER be an object returned from Hint corresponding to a specific biomedical\n",
      "                            entity, OR be a string or list of strings corresponding to Biolink Entity classes.\n",
      "                            Examples:\n",
      "                Hint().query(\"acetaminophen\")['ChemicalSubstance'][0]\n",
      "                'Gene'\n",
      "                ['Gene','ChemicalSubstance']\n",
      "    \n",
      "        intermediate_nodes (required): the semantic type(s) of the intermediate node(s).  Examples:\n",
      "                None                         : no intermediate node, find direct connections only\n",
      "                []                           : no intermediate node, find direct connections only\n",
      "                ['BiologicalEntity']         : one intermediate node of any semantic type\n",
      "                ['Gene']                     : one intermediate node that must be a Gene\n",
      "                [('Gene','Pathway')]         : one intermediate node that must be a Gene or a Pathway\n",
      "                ['Gene','Pathway']           : two intermediate nodes, first must be a Gene, second must be a Pathway.\n",
      "                ['Gene',('Pathway','Gene')]  : two intermediate nodes, first must be a Gene, second must be a Pathway or Gene.\n",
      "                                                **NOTE**: queries with more than one intermediate node are currently not supported\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(FindConnection.__init__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we formulate a `FindConnection` query with \"CML\" as the `input_ojb`, \"imatinib\" as the `output_obj`.  We further specify with the `intermediate_nodes` parameter that we are looking for paths joining chronic myelogenous leukemia and imatinib with *one* intermediate node that is a Gene.  (The ability to search for longer reasoning paths that include additional intermediate nodes will be added shortly.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc = FindConnection(input_obj=cml, output_obj=imatinib, intermediate_nodes='Gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We next execute the `connect` method, which performs the **query path planning** and **query path execution** process.  In short, BioThings Explorer is deconstructing the query into individual API calls, executing those API calls, then assembling the results.\n",
    "\n",
    "A verbose log of this process is displayed below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==========\n",
      "========== QUERY PARAMETER SUMMARY ==========\n",
      "==========\n",
      "\n",
      "BTE will find paths that join 'chronic myelogenous leukemia' and 'Imatinib'. Paths will have 1 intermediate node.\n",
      "\n",
      "Intermediate node #1 will have these type constraints: Gene\n",
      "\n",
      "\n",
      "==========\n",
      "========== QUERY #1 -- fetch all Gene entities linked to 'chronic myelogenous leukemia' ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because chronic myelogenous leukemia is of type 'DiseaseOrPhenotypicFeature', BTE will query our meta-KG for APIs that can take 'DiseaseOrPhenotypicFeature' as input and 'Gene' as output\n",
      "\n",
      "BTE found 7 apis:\n",
      "\n",
      "API 1. mydisease.info(1 API call)\n",
      "API 2. DISEASES(1 API call)\n",
      "API 3. biolink_disease2gene(1 API call)\n",
      "API 4. mgigene2phenotype(1 API call)\n",
      "API 5. scibite_disease2gene(1 API call)\n",
      "API 6. semmeddisease(1 API call)\n",
      "API 7. corddisease(1 API call)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 7.1: https://biothings.ncats.io/cord_disease/query (POST \"q=DOID:8552&scopes=doid&fields=associated_with&species=human&size=100\")\n",
      "API 2.1: https://biothings.ncats.io/DISEASES/query (POST \"q=MONDO:0011996&scopes=_id&fields=DISEASES.associatedWith&species=human&size=100\")\n",
      "API 4.1: https://biothings.ncats.io/mgigene2phenotype/query (POST \"q=DOID:8552&scopes=mgi.associated_with_disease.doid&fields=_id&species=human&size=100\")\n",
      "API 3.1: https://api.monarchinitiative.org/api/bioentity/disease/MONDO:0011996/genes?rows=100\n",
      "API 1.1: http://mydisease.info/v1/query (POST \"q=C1292772,C0023473&scopes=mondo.xrefs.umls,disgenet.xrefs.umls&fields=disgenet.genes_related_to_disease&species=human&size=100\")\n",
      "API 5.1: https://automat.renci.org/cord19_scibite_v2/disease/gene/MONDO:0011996\n",
      "API 6.1: http://biothings.ncats.io/semmed/query (POST \"q=C1292772,C0023473&scopes=umls&fields=physically_interacts_with,negatively_regulates,positively_regulated_by,disrupts,prevented_by,produced_by,preceded_by,subclass_of,affected_by,positively_regulates,negatively_regulated_by,derives_from,treated_by,related_to,predisposed_by,part_of,caused_by,has_subclass,coexists_with,precedes,affects,disrupted_by,has_part,manifestation_of,produces&species=human&size=100\")\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 2.1 DISEASES: 3 hits\n",
      "API 3.1 biolink_disease2gene: 4 hits\n",
      "API 7.1 corddisease: 146 hits\n",
      "API 4.1 mgigene2phenotype: 8 hits\n",
      "API 1.1 mydisease.info: 52 hits\n",
      "API 5.1 scibite_disease2gene: 1 hits\n",
      "API 6.1 semmeddisease: 1139 hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 842 unique objects.\n",
      "\n",
      "\n",
      "==========\n",
      "========== QUERY #2 -- fetch all Gene entities linked to 'Imatinib' ==========\n",
      "==========\n",
      "\n",
      "==== Step #1: Query path planning ====\n",
      "\n",
      "Because Imatinib is of type 'ChemicalSubstance', BTE will query our meta-KG for APIs that can take 'ChemicalSubstance' as input and 'Gene' as output\n",
      "\n",
      "BTE found 7 apis:\n",
      "\n",
      "API 1. dgidb_chemical2gene(1 API call)\n",
      "API 2. scibite_chemical2gene(1 API call)\n",
      "API 3. ctd_chemical2gene(1 API call)\n",
      "API 4. chembl_drug_mechanism(1 API call)\n",
      "API 5. semmedchemical(1 API call)\n",
      "API 6. cordchemical(1 API call)\n",
      "API 7. mychem.info(2 API calls)\n",
      "\n",
      "\n",
      "==== Step #2: Query path execution ====\n",
      "NOTE: API requests are dispatched in parallel, so the list of APIs below is ordered by query time.\n",
      "\n",
      "API 5.1: https://biothings.ncats.io/cord_chemical/query (POST \"q=CHEBI:45783&scopes=chebi&fields=associated_with&species=human&size=100\")\n",
      "API 4.1: https://biothings.ncats.io/semmedchemical/query (POST \"q=C0935989&scopes=umls&fields=physically_interacts_with,affects,derives_info,subclass_of,negatively_regulates,affected_by,positively_regulated_by,disrupts,negatively_regulated_by,derives_from,part_of,positively_regulates,disrupted_by,has_part,produced_by,related_to,has_subclass,coexists_with,precedes,produces&species=human&size=100\")\n",
      "API 7.1: https://www.ebi.ac.uk/chembl/api/data/mechanism.json?molecule_chembl_id=CHEMBL941\n",
      "API 1.1: http://www.dgidb.org/api/v2/interactions.json?drugs=CHEMBL941\n",
      "scibite_chemical2gene failed\n",
      "API 6.2: http://mychem.info/v1/query (POST \"q=DB00619&scopes=drugbank.id&fields=drugbank.enzymes,drugbank.targets&species=human&size=100\")\n",
      "API 6.1: http://mychem.info/v1/query (POST \"q=CHEMBL941&scopes=chembl.molecule_chembl_id,chebi.xrefs.chembl&fields=drugcentral.bioactivity&species=human&size=100\")\n",
      "API 3.1: http://ctdbase.org/tools/batchQuery.go?inputType=chem&inputTerms=D000068877|mercury&report=genes_curated&format=json\n",
      "0, message='Attempt to decode JSON with unexpected mimetype: text/plain;charset=utf-8', url='http://ctdbase.org/tools/batchQuery.go?inputType=chem&inputTerms=D000068877%7Cmercury&report=genes_curated&format=json\n",
      "\n",
      "\n",
      "==== Step #3: Output normalization ====\n",
      "\n",
      "API 7.1 chembl_drug_mechanism: No hits\n",
      "API 5.1 cordchemical: 172 hits\n",
      "API 3.1 ctd_chemical2gene: 910 hits\n",
      "API 1.1 dgidb_chemical2gene: 34 hits\n",
      "API 6.1 mychem.info: 77 hits\n",
      "API 6.2 mychem.info: 18 hits\n",
      "API 2.1 scibite_chemical2gene: No hits\n",
      "API 4.1 semmedchemical: 802 hits\n",
      "\n",
      "After id-to-object translation, BTE retrieved 1298 unique objects.\n",
      "\n",
      "==========\n",
      "========== Final assembly of results ==========\n",
      "==========\n",
      "\n",
      "\n",
      "BTE found 300 unique intermediate nodes connecting 'chronic myelogenous leukemia' and 'Imatinib'\n"
     ]
    }
   ],
   "source": [
    "# set verbose=True will display all steps which BTE takes to find the connection\n",
    "fc.connect(verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3: Display and Filter results\n",
    "This section demonstrates post-query filtering done in Python. Later, more advanced filtering functions will be added to the **query path execution** module for interleaved filtering, thereby enabling longer query paths. More details to come...\n",
    "\n",
    "First, all matching paths can be exported to a data frame. Let's examine a sample of those results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>input_type</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred1_source</th>\n",
       "      <th>pred1_api</th>\n",
       "      <th>pred1_pubmed</th>\n",
       "      <th>node1_type</th>\n",
       "      <th>node1_name</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>pred2</th>\n",
       "      <th>pred2_source</th>\n",
       "      <th>pred2_api</th>\n",
       "      <th>pred2_pubmed</th>\n",
       "      <th>output_type</th>\n",
       "      <th>output_name</th>\n",
       "      <th>output_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>132</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>DISEASES</td>\n",
       "      <td>DISEASES</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>entrez:25</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>ctd</td>\n",
       "      <td>ctd_chemical2gene</td>\n",
       "      <td>12446442,15039284,15107311,15329907,16849647,1...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>621</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>CORD API</td>\n",
       "      <td>corddisease</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BCR</td>\n",
       "      <td>entrez:613</td>\n",
       "      <td>target</td>\n",
       "      <td>drugbank</td>\n",
       "      <td>mychem.info</td>\n",
       "      <td>15206509,20072827,12600228,23226582</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1798</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>CORD API</td>\n",
       "      <td>corddisease</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>INS</td>\n",
       "      <td>entrez:3630</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>CORD API</td>\n",
       "      <td>cordchemical</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>800</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td>2835901,11766349,12218235,12145676,12145676,12...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BCR</td>\n",
       "      <td>entrez:613</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmedchemical</td>\n",
       "      <td>22372463</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>857</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>predisposed_by</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td>21039054</td>\n",
       "      <td>Gene</td>\n",
       "      <td>CYP3A5</td>\n",
       "      <td>entrez:1577</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmedchemical</td>\n",
       "      <td>23394475</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2445</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td>10590068,18202009</td>\n",
       "      <td>Gene</td>\n",
       "      <td>CXCR4</td>\n",
       "      <td>entrez:7852</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmedchemical</td>\n",
       "      <td>18202009,18202009,18202009</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2221</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td>2098290,2098290</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IL2</td>\n",
       "      <td>entrez:3558</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>ctd</td>\n",
       "      <td>ctd_chemical2gene</td>\n",
       "      <td>12730625</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2230</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>related_to</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td>24118457</td>\n",
       "      <td>Gene</td>\n",
       "      <td>MAFF</td>\n",
       "      <td>entrez:23764</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>ctd</td>\n",
       "      <td>ctd_chemical2gene</td>\n",
       "      <td>19937285</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>613</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td>16778200,18203007,21714205,19878872,20194843,2...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>BCR</td>\n",
       "      <td>entrez:613</td>\n",
       "      <td>target</td>\n",
       "      <td>drugcentral</td>\n",
       "      <td>mychem.info</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2273</td>\n",
       "      <td>chronic myelogenous leukemia</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>has_part</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmeddisease</td>\n",
       "      <td>11146162,11146162,9746786,11756187,10973491,10...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>CD34</td>\n",
       "      <td>entrez:947</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmedchemical</td>\n",
       "      <td>20621681,22829973</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             input                  input_type  \\\n",
       "132   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature   \n",
       "621   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature   \n",
       "1798  chronic myelogenous leukemia  DiseaseOrPhenotypicFeature   \n",
       "800   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature   \n",
       "857   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature   \n",
       "2445  chronic myelogenous leukemia  DiseaseOrPhenotypicFeature   \n",
       "2221  chronic myelogenous leukemia  DiseaseOrPhenotypicFeature   \n",
       "2230  chronic myelogenous leukemia  DiseaseOrPhenotypicFeature   \n",
       "613   chronic myelogenous leukemia  DiseaseOrPhenotypicFeature   \n",
       "2273  chronic myelogenous leukemia  DiseaseOrPhenotypicFeature   \n",
       "\n",
       "                pred1 pred1_source      pred1_api  \\\n",
       "132    associatedWith     DISEASES       DISEASES   \n",
       "621   associated_with     CORD API    corddisease   \n",
       "1798  associated_with     CORD API    corddisease   \n",
       "800        treated_by   SEMMED API  semmeddisease   \n",
       "857    predisposed_by   SEMMED API  semmeddisease   \n",
       "2445       related_to   SEMMED API  semmeddisease   \n",
       "2221       treated_by   SEMMED API  semmeddisease   \n",
       "2230       related_to   SEMMED API  semmeddisease   \n",
       "613      disrupted_by   SEMMED API  semmeddisease   \n",
       "2273         has_part   SEMMED API  semmeddisease   \n",
       "\n",
       "                                           pred1_pubmed node1_type node1_name  \\\n",
       "132                                                None       Gene       ABL1   \n",
       "621                                                None       Gene        BCR   \n",
       "1798                                               None       Gene        INS   \n",
       "800   2835901,11766349,12218235,12145676,12145676,12...       Gene        BCR   \n",
       "857                                            21039054       Gene     CYP3A5   \n",
       "2445                                  10590068,18202009       Gene      CXCR4   \n",
       "2221                                    2098290,2098290       Gene        IL2   \n",
       "2230                                           24118457       Gene       MAFF   \n",
       "613   16778200,18203007,21714205,19878872,20194843,2...       Gene        BCR   \n",
       "2273  11146162,11146162,9746786,11756187,10973491,10...       Gene       CD34   \n",
       "\n",
       "          node1_id                      pred2 pred2_source          pred2_api  \\\n",
       "132      entrez:25             associatedWith          ctd  ctd_chemical2gene   \n",
       "621     entrez:613                     target     drugbank        mychem.info   \n",
       "1798   entrez:3630            associated_with     CORD API       cordchemical   \n",
       "800     entrez:613  physically_interacts_with   SEMMED API     semmedchemical   \n",
       "857    entrez:1577              coexists_with   SEMMED API     semmedchemical   \n",
       "2445   entrez:7852       positively_regulates   SEMMED API     semmedchemical   \n",
       "2221   entrez:3558             associatedWith          ctd  ctd_chemical2gene   \n",
       "2230  entrez:23764             associatedWith          ctd  ctd_chemical2gene   \n",
       "613     entrez:613                     target  drugcentral        mychem.info   \n",
       "2273    entrez:947  physically_interacts_with   SEMMED API     semmedchemical   \n",
       "\n",
       "                                           pred2_pubmed output_type  \\\n",
       "132   12446442,15039284,15107311,15329907,16849647,1...        Gene   \n",
       "621                 15206509,20072827,12600228,23226582        Gene   \n",
       "1798                                               None        Gene   \n",
       "800                                            22372463        Gene   \n",
       "857                                            23394475        Gene   \n",
       "2445                         18202009,18202009,18202009        Gene   \n",
       "2221                                           12730625        Gene   \n",
       "2230                                           19937285        Gene   \n",
       "613                                                None        Gene   \n",
       "2273                                  20621681,22829973        Gene   \n",
       "\n",
       "     output_name         output_id  \n",
       "132     IMATINIB  chembl:CHEMBL941  \n",
       "621     IMATINIB  chembl:CHEMBL941  \n",
       "1798    IMATINIB  chembl:CHEMBL941  \n",
       "800     IMATINIB  chembl:CHEMBL941  \n",
       "857     IMATINIB  chembl:CHEMBL941  \n",
       "2445    IMATINIB  chembl:CHEMBL941  \n",
       "2221    IMATINIB  chembl:CHEMBL941  \n",
       "2230    IMATINIB  chembl:CHEMBL941  \n",
       "613     IMATINIB  chembl:CHEMBL941  \n",
       "2273    IMATINIB  chembl:CHEMBL941  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fc.display_table_view()\n",
    "\n",
    "# because UMLS is not currently well-integrated in our ID-to-object translation system, removing UMLS-only entries here\n",
    "patternDel = \"^umls:C\\d+\"\n",
    "filter = df.node1_id.str.contains(patternDel)\n",
    "df = df[~filter]\n",
    "\n",
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While most results are based on edges from [semmed](https://skr3.nlm.nih.gov/SemMed/), edges from [DGIdb](http://www.dgidb.org/), [biolink](https://monarchinitiative.org/), [disgenet](http://www.disgenet.org/), [mydisease.info](https://mydisease.info) and [drugcentral](http://drugcentral.org/) were also retrieved from their respective APIs.  \n",
    "\n",
    "Next, let's look to see which genes are mentioned the most."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BCR       221\n",
       "ABL1      156\n",
       "KIT       130\n",
       "ABCB1      75\n",
       "MTTP       63\n",
       "AKT1       55\n",
       "TP53       49\n",
       "CD34       36\n",
       "TNF        32\n",
       "STAT5A     28\n",
       "Name: node1_name, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.node1_name.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not surprisingly, the top two genes that BioThings Explorer found that join imatinib to CML are *ABL1* and *BCR*, the two genes that are fused in the \"Philadelphia chromosome\", the genetic abnormality that underlies CML, *and* the validate target of imatinib.\n",
    "\n",
    "Let's examine some of the PubMed articles linking **CML to *ABL1*** and ***ABL1* to imatinib**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 22 articles supporting the edge between CML and ABL1. Sampling of 10 of those:\n",
      "http://pubmed.gov/18243808\n",
      "http://pubmed.gov/25389112\n",
      "http://pubmed.gov/26186983\n",
      "http://pubmed.gov/27282582\n",
      "http://pubmed.gov/23955590\n",
      "http://pubmed.gov/27904889\n",
      "http://pubmed.gov/11979553\n",
      "http://pubmed.gov/10498618\n",
      "http://pubmed.gov/10822991\n",
      "http://pubmed.gov/11368359\n"
     ]
    }
   ],
   "source": [
    "# fetch all articles connecting 'chronic myelogenous leukemia' and 'ABL1'\n",
    "articles = []\n",
    "for info in fc.display_edge_info('chronic myelogenous leukemia', 'ABL1').values():\n",
    "    if 'pubmed' in info['info']:\n",
    "        articles += info['info']['pubmed']\n",
    "print(\"There are \"+str(len(articles))+\" articles supporting the edge between CML and ABL1. Sampling of 10 of those:\")\n",
    "x = [print(\"http://pubmed.gov/\"+str(x)) for x in articles[0:10] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 33 articles supporting the edge between ABL1 and imatinib. Sampling of 10 of those:\n",
      "http://pubmed.gov/15799618\n",
      "http://pubmed.gov/15917650\n",
      "http://pubmed.gov/15949566\n",
      "http://pubmed.gov/16153117\n",
      "http://pubmed.gov/16205964\n",
      "http://pubmed.gov/12032135\n",
      "http://pubmed.gov/12446442\n",
      "http://pubmed.gov/15039284\n",
      "http://pubmed.gov/15107311\n",
      "http://pubmed.gov/15329907\n"
     ]
    }
   ],
   "source": [
    "# fetch all articles connecting 'ABL1' and 'Imatinib\n",
    "articles = []\n",
    "for info in fc.display_edge_info('ABL1', 'Imatinib').values():\n",
    "    if 'pubmed' in info['info']:\n",
    "        articles += info['info']['pubmed']\n",
    "print(\"There are \"+str(len(articles))+\" articles supporting the edge between ABL1 and imatinib. Sampling of 10 of those:\")\n",
    "x = [print(\"http://pubmed.gov/\"+str(x)) for x in articles[0:10] ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing results between CML and GIST\n",
    "\n",
    "Let's perform another BioThings Explorer query, this time looking to EXPLAIN the relationship between imatinib and gastrointestinal stromal tumors (GIST), another disease treated by imatinib."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mondo': 'MONDO:0011719',\n",
       " 'doid': 'DOID:9253',\n",
       " 'umls': 'C3179349',\n",
       " 'mesh': 'D046152',\n",
       " 'name': 'gastrointestinal stromal tumor',\n",
       " 'display': 'mondo(MONDO:0011719) doid(DOID:9253) umls(C3179349) mesh(D046152) name(gastrointestinal stromal tumor) ',\n",
       " 'type': 'DiseaseOrPhenotypicFeature',\n",
       " 'primary': {'identifier': 'mondo',\n",
       "  'cls': 'DiseaseOrPhenotypicFeature',\n",
       "  'value': 'MONDO:0011719'}}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ht = Hint()\n",
    "# find all potential representations of CML\n",
    "gist_hint = ht.query(\"gastrointestinal stromal tumor\")\n",
    "# select the correct representation of CML\n",
    "gist = gist_hint['DiseaseOrPhenotypicFeature'][0]\n",
    "gist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc = FindConnection(input_obj=gist, output_obj=imatinib, intermediate_nodes='Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "fc.connect(verbose=False) # skipping the verbose log here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input</th>\n",
       "      <th>input_type</th>\n",
       "      <th>pred1</th>\n",
       "      <th>pred1_source</th>\n",
       "      <th>pred1_api</th>\n",
       "      <th>pred1_pubmed</th>\n",
       "      <th>node1_type</th>\n",
       "      <th>node1_name</th>\n",
       "      <th>node1_id</th>\n",
       "      <th>pred2</th>\n",
       "      <th>pred2_source</th>\n",
       "      <th>pred2_api</th>\n",
       "      <th>pred2_pubmed</th>\n",
       "      <th>output_type</th>\n",
       "      <th>output_name</th>\n",
       "      <th>output_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>152</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>scibite</td>\n",
       "      <td>scibite_disease2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>entrez:3815</td>\n",
       "      <td>target</td>\n",
       "      <td>dgidb</td>\n",
       "      <td>dgidb_chemical2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>145</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>scibite</td>\n",
       "      <td>scibite_disease2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>entrez:3815</td>\n",
       "      <td>target</td>\n",
       "      <td>drugbank</td>\n",
       "      <td>mychem.info</td>\n",
       "      <td>16865565,16087693,17458563,17369583,17559139</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>148</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>DISEASES</td>\n",
       "      <td>DISEASES</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>entrez:3815</td>\n",
       "      <td>target</td>\n",
       "      <td>dgidb</td>\n",
       "      <td>dgidb_chemical2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>162</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>DISEASES</td>\n",
       "      <td>DISEASES</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>entrez:3815</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>CORD API</td>\n",
       "      <td>cordchemical</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>375</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>CORD API</td>\n",
       "      <td>corddisease</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>TP53</td>\n",
       "      <td>entrez:7157</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmedchemical</td>\n",
       "      <td>20876800</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>180</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>scibite</td>\n",
       "      <td>scibite_disease2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>entrez:3815</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmedchemical</td>\n",
       "      <td>16143881,17208434,17372901,18234488,21442641,2...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>173</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>scibite</td>\n",
       "      <td>scibite_disease2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>entrez:3815</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmedchemical</td>\n",
       "      <td>24369323,24369323,28334365</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>50</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>DISEASES</td>\n",
       "      <td>DISEASES</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>entrez:5156</td>\n",
       "      <td>target</td>\n",
       "      <td>dgidb</td>\n",
       "      <td>dgidb_chemical2gene</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>291</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>CORD API</td>\n",
       "      <td>corddisease</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>KIT</td>\n",
       "      <td>entrez:3815</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>SEMMED API</td>\n",
       "      <td>semmedchemical</td>\n",
       "      <td>23637779</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>32</td>\n",
       "      <td>gastrointestinal stromal tumor</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>associatedWith</td>\n",
       "      <td>DISEASES</td>\n",
       "      <td>DISEASES</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>entrez:5156</td>\n",
       "      <td>target</td>\n",
       "      <td>drugcentral</td>\n",
       "      <td>mychem.info</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene</td>\n",
       "      <td>IMATINIB</td>\n",
       "      <td>chembl:CHEMBL941</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              input                  input_type  \\\n",
       "152  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "145  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "148  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "162  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "375  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "180  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "173  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "50   gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "291  gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "32   gastrointestinal stromal tumor  DiseaseOrPhenotypicFeature   \n",
       "\n",
       "               pred1 pred1_source             pred1_api pred1_pubmed  \\\n",
       "152   associatedWith      scibite  scibite_disease2gene         None   \n",
       "145   associatedWith      scibite  scibite_disease2gene         None   \n",
       "148   associatedWith     DISEASES              DISEASES         None   \n",
       "162   associatedWith     DISEASES              DISEASES         None   \n",
       "375  associated_with     CORD API           corddisease         None   \n",
       "180   associatedWith      scibite  scibite_disease2gene         None   \n",
       "173   associatedWith      scibite  scibite_disease2gene         None   \n",
       "50    associatedWith     DISEASES              DISEASES         None   \n",
       "291  associated_with     CORD API           corddisease         None   \n",
       "32    associatedWith     DISEASES              DISEASES         None   \n",
       "\n",
       "    node1_type node1_name     node1_id                      pred2  \\\n",
       "152       Gene        KIT  entrez:3815                     target   \n",
       "145       Gene        KIT  entrez:3815                     target   \n",
       "148       Gene        KIT  entrez:3815                     target   \n",
       "162       Gene        KIT  entrez:3815            associated_with   \n",
       "375       Gene       TP53  entrez:7157  physically_interacts_with   \n",
       "180       Gene        KIT  entrez:3815              coexists_with   \n",
       "173       Gene        KIT  entrez:3815              coexists_with   \n",
       "50        Gene     PDGFRA  entrez:5156                     target   \n",
       "291       Gene        KIT  entrez:3815  physically_interacts_with   \n",
       "32        Gene     PDGFRA  entrez:5156                     target   \n",
       "\n",
       "    pred2_source            pred2_api  \\\n",
       "152        dgidb  dgidb_chemical2gene   \n",
       "145     drugbank          mychem.info   \n",
       "148        dgidb  dgidb_chemical2gene   \n",
       "162     CORD API         cordchemical   \n",
       "375   SEMMED API       semmedchemical   \n",
       "180   SEMMED API       semmedchemical   \n",
       "173   SEMMED API       semmedchemical   \n",
       "50         dgidb  dgidb_chemical2gene   \n",
       "291   SEMMED API       semmedchemical   \n",
       "32   drugcentral          mychem.info   \n",
       "\n",
       "                                          pred2_pubmed output_type  \\\n",
       "152                                               None        Gene   \n",
       "145       16865565,16087693,17458563,17369583,17559139        Gene   \n",
       "148                                               None        Gene   \n",
       "162                                               None        Gene   \n",
       "375                                           20876800        Gene   \n",
       "180  16143881,17208434,17372901,18234488,21442641,2...        Gene   \n",
       "173                         24369323,24369323,28334365        Gene   \n",
       "50                                                None        Gene   \n",
       "291                                           23637779        Gene   \n",
       "32                                                None        Gene   \n",
       "\n",
       "    output_name         output_id  \n",
       "152    IMATINIB  chembl:CHEMBL941  \n",
       "145    IMATINIB  chembl:CHEMBL941  \n",
       "148    IMATINIB  chembl:CHEMBL941  \n",
       "162    IMATINIB  chembl:CHEMBL941  \n",
       "375    IMATINIB  chembl:CHEMBL941  \n",
       "180    IMATINIB  chembl:CHEMBL941  \n",
       "173    IMATINIB  chembl:CHEMBL941  \n",
       "50     IMATINIB  chembl:CHEMBL941  \n",
       "291    IMATINIB  chembl:CHEMBL941  \n",
       "32     IMATINIB  chembl:CHEMBL941  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fc.display_table_view()\n",
    "\n",
    "# because UMLS is not currently well-integrated in our ID-to-object translation system, removing UMLS-only entries here\n",
    "patternDel = \"^umls:C\\d+\"\n",
    "filter = df.node1_id.str.contains(patternDel)\n",
    "df = df[~filter]\n",
    "\n",
    "df.sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KIT       120\n",
       "PDGFRA     52\n",
       "ABL1       18\n",
       "BCR        15\n",
       "TP53       14\n",
       "FIP1L1     12\n",
       "MTTP        9\n",
       "BRAF        8\n",
       "CD34        6\n",
       "EGFR        6\n",
       "Name: node1_name, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.node1_name.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, the top two genes that BioThings Explorer found that join imatinib to GIST are *PDGFRA* and *KIT*, the most commonly mutated genes found in GIST *and* validated targets of imatinib.\n",
    "\n",
    "While several of the listed genes would be considered positive controls, others on the list could be viewed as **testable hypotheses and discovery opportunities** to be evaluated by domain experts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions and caveats\n",
    "\n",
    "This notebook demonstrated the use of BioThings Explorer in EXPLAIN mode to investigate the relationship between imatinib and two diseases that it treats -- chronic myelogenous leukemia (CML) and gastrointestinal stromal tumors (GIST).  In each case, BioThings Explorer autonomously queried a **distributed knowledge graph of biomedical APIs** to find the most common genes, and in each case the relevant targets were retrieved.\n",
    "\n",
    "There are still many areas for improvement (and some areas in which BioThings Explorer is still buggy).  And of course, BioThings Explorer is dependent on the accessibility of the APIs that comprise the distributed knowledge graph.  Nevertheless, we encourage users to try other variants of the EXPLAIN queries demonstrated in this notebook."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
